### PR TITLE
Support for PyMongo 3.10 and removal of nose

### PR DESCRIPTION
**Pull request by amol-**

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
       include_package_data=True,
       zip_safe=True,
       install_requires=[
-        "pymongo>=3.0,<3.8",
+        "pymongo>=3.0,<3.12",
```
Comment (brondsem) [inline setup.py:38]: If we only want to allow up to 3.10, I think it should be:
```suggestion
        "pymongo>=3.0,<3.11",
```

**Comment by brondsem** [comment]
I'm also getting the following error.  I bisected it to 8ea40d7 `Make order in mappers registration` 
```
  vi +40  allura/websetup/__init__.py  # setup_app
    bootstrap.bootstrap(command, conf, vars)
  vi +96  allura/websetup/bootstrap.py  # bootstrap
    display_name='Anonymous')
  vi +405 /Users/brondsem/sf/ming/ming/odm/mapper.py  # __init__
    self.func(self_, *args, **kwargs)
TypeError: __init__() got an unexpected keyword argument '_id'
```

The bootstrap.py lines look like:
```
    M.User(
        _id=None,
        username='*anonymous',
        display_name='Anonymous')
```
and the `User` class inherits directly from `MappedClass` and some mixins.  It has no `__init__` method

**Comment by amol-** [comment]
Thanks for catching those, Ming test suite is definitely not super effective and I didn't have a chance to test things with a real world project yet. I'll check what's going on.

**Comment by amol-** [comment]
> 
> and the `User` class inherits directly from `MappedClass` and some mixins. It has no `__init__` method

Curiously I tried to replicate the scenario in https://github.com/TurboGears/Ming/pull/18/commits/ae7ea31d743e5a7c0d90c7f1f28b7de71ed2ed28 but it has been passing fine.

**Comment by amol-** [comment]
Ok, I think that the mixins might be involved and we might have ended using the init of one of the mixins. 

```
class User(MappedClass, ActivityNode, ActivityObject, SearchIndexable):
```


In allura the `ActivityNode` inherits from `activitystream.base.NodeBase` which has an `__init__` accepting no arguments.

**Comment by amol-** [comment]
Yes, confirmed that adding an `__init__` to the mixins in the test case reproduces the same exact error. I feel that the Mixins are not properly written as they should proxy the init arguments to super etc... but we should be able to restore the previous behaviour.

**Comment by brondsem** [comment]
I agree, fixes in both places sounds best

**Comment by amol-** [comment]
I restored the previous behaviour of `MappedClass` providing a custom init (`_basic_init`) that should address the issue with the Mixin providing its own `__init__`.

**Comment by brondsem** [comment]
That helps, thanks.  There actually were quite a few places in Allura that had `__init__` problems even if I changed the ActivityNode mixin.  FYI here's a very simple example, without even a mixin:

```
class ProjectRole(MappedClass):
    ...
    def __init__(self, **kw):
        assert 'project_id' in kw, 'Project roles must specify a project id'
        super(ProjectRole, self).__init__(**kw)
```
errors with:
`TypeError: object.__init__() takes no parameters`

The latest commit does fix that, but wanted to share it in case it is helpful.

There are still some other inheritance problems, though.  At least the `delete()` method has a problem, so I wrote a test for it:

```
    def test_delete_super(self):

        class User(MappedClass):
            class __mongometa__:
                name = "user_with_custom_delete"
                session = self.session

            _id = FieldProperty(S.ObjectId)
            username = FieldProperty(str)
            
            def delete(self):
                super(User, self).delete()

        u = User(_id=None, username="anonymous")
        self.session.flush()
        u.delete()
        self.session.flush()

        assert not User.query.find({"username": "anonymous"}).count()
```
fails with this error:
```
  File "/Users/brondsem/sf/ming/ming/tests/odm/test_declarative.py", line 81, in delete
    super(User, self).delete()
AttributeError: 'super' object has no attribute 'delete'
```

**Comment by amol-** [comment]
Ok, I gave up trying to exclude `MappedClass` from the registered mappers. I find confusing that it figures as one as it can't be used to query anything, but it seems it leads to a much more complex code than just letting it be.

I restored the previous behaviour on MappedClass and added your test for `super.delete` which passes fine.

**Comment by pgmustard-admin** [comment]
Hi, 
Is there going to be a new release with this change incorporated soon?

Thanks

**Comment by brondsem** [comment]
> Hi,
> Is there going to be a new release with this change incorporated soon?
> 
> Thanks

@amol- what do you think?  I can cut a release if that helps.  I do have a few merge requests I'd like to have released as well.  I am fairly confident about those merge requests as well as other recent changes.  I've been able to test Allura and an internal app successfully with latest changes + merge requests.

**Comment by amol-** [comment]
I made a 0.10 release. with the changes that were safe. My plan is to now also merge the other two PRs and then incorporate them in a subsequent release.

**Pull request merged**